task_sdk/src/airflow/sdk/execution_time/supervisor.py

```diff
@@ -403,12 +403,57 @@ def _send_startup_message(self, ti: TaskInstance, path: str | os.PathLike[str],
         self.stdin.write(msg.model_dump_json().encode())
         self.stdin.write(b"\n")
 
-    def kill(self, signal: signal.Signals = signal.SIGINT):
+    def kill(
+        self,
+        signal_to_send: signal.Signals = signal.SIGINT,
+        escalation_delay: float = 5.0,
+        force: bool = False,
+    ):
+        """
+        Attempt to terminate the subprocess with a given signal.
+
+        If the process does not exit within `escalation_delay` seconds, escalate to SIGTERM and eventually SIGKILL if necessary.
+
+        :param signal_to_send: The signal to send initially (default is SIGINT).
+        :param escalation_delay: Time in seconds to wait before escalating to a stronger signal.
+        :param force: If True, send the signal immediately without escalation.
+        """
         if self._exit_code is not None:
             return
 
-        with suppress(ProcessLookupError):
-            os.kill(self.pid, signal)
+        if not force:
+            with suppress(ProcessLookupError):
+                self._process.send_signal(signal_to_send)
+            return
+
+        # Escalation sequence: SIGINT -> SIGTERM -> SIGKILL
+        escalation_path = [signal.SIGINT, signal.SIGTERM, signal.SIGKILL]
+        if signal_to_send in escalation_path:
+            # Start from `initial_signal`
+            escalation_path = escalation_path[escalation_path.index(signal_to_send) :]
+
+        for sig in escalation_path:
+            try:
+                self._process.send_signal(sig)
```

Comment:
Following discussion we had in https://github.com/apache/airflow/pull/41329#issuecomment-2506804789. ->  should we create a new process group when we start task via supervisor and send the signal to process group ? 

This handles way better situation where one of the child processes does not propagate the signals (which happens for example by default in bash).

Typical approach here is to start a new process group with same id as the process it starts (setpgid(0,0)) right after we fork from supervisor:

```python
        # Check if we are not a group leader already (We should not be)
        if os.getpid() != os.getsid(0):
            # and create a new process group where we are the leader
            os.setpgid(0, 0)
```

And then instead of sending signal to process you send it to process group

```python
 os.killpg(gid, signal.SIGTERM)
```

Reply:
This is for example what Ctrl-C does by default when you have interactive process - the TERM signal is sent to the foreground process group that shell sets for the foreground process - not **just** to the process.

Reply:
Yes we should create a new process group, there is already a todo for that, so that will be in a different PR.

Reply:
Though we should try and support non-UNIX (i.e. Windows) which doesn't have process groups.

Reply:
> Though we should try and support non-UNIX (i.e. Windows) which doesn't have process groups.

SIGTERM does not work on Windows either, so we will have to do it completely differently if we want to support Windows. 

See for example here: https://stackoverflow.com/questions/47306805/signal-sigterm-not-received-by-subprocess-on-windows

Reply:
Not mentioning fork that also deoes not **really** work well on MacOS (because on MacOS system libraries might start threads which does not play well with threads https://github.com/python/cpython/issues/77906), so if we would like to have explicit support for MacOS (and later Windows) for "workers" -  we will have to make some serious ifs

BTW. In Python 3.14 there are some changes coming that **might** finally result in a possibility of using fork() in production for MacOS https://docs.python.org/3/library/multiprocessing.html#multiprocessing-start-methods - maybe we should already add "get_context()" and "set_start_method" to explicity set "fork" on systems that support it (minus MacOS :)  - as this will be a breaking change in Python 3.14

Reply:
https://github.com/giampaolo/psutil/blob/master/psutil/__init__.py#L1278-L1291

psutil manages some of that for us:

```py
    def send_signal(self, sig):
        """Send a signal *sig* to process pre-emptively checking
        whether PID has been reused (see signal module constants) .
        On Windows only SIGTERM is valid and is treated as an alias
        for kill().
        """
        if POSIX:
            self._send_signal(sig)
        else:  # pragma: no cover
            self._raise_if_pid_reused()
            if sig != signal.SIGTERM and not self.is_running():
                msg = "process no longer exists"
                raise NoSuchProcess(self.pid, self._name, msg=msg)
            self._proc.send_signal(sig)
```

Reply:
interesting :)